src/Abstractions/NexusMods.Abstractions.Loadouts.Synchronizers/ALoadoutSynchronizer.cs

```diff
@@ -967,26 +983,51 @@ public async Task IndexNewState(GameInstallation installation, ITransaction tx)
 
             await Parallel.ForEachAsync(location.Value.EnumerateFiles(), async (file, token) =>
                 {
-                    var gamePath = installation.LocationsRegister.ToGamePath(file);
-                    if (IsIgnoredPath(gamePath))
-                    {
-                        return;
-                    }
-
-                    var newHash = await file.XxHash3Async(token: token);
+                    var (gamePath, newHash) = await HashGameFile(hashDb, installation, file, token);
                     _ = new DiskStateEntry.New(tx, tx.TempId(DiskStateEntry.EntryPartition))
                     {
                         Path = gamePath.ToGamePathParentTuple(metaDataId),
                         Hash = newHash,
                         Size = file.FileInfo.Size,
                         LastModified = file.FileInfo.LastWriteTimeUtc,
-                        GameId = metaDataId
+                        GameId = metaDataId,
                     };
                 }
             );
         }
     }
 
+    /// <summary>
+    /// Hash a file but first check the hash database for a matching size and minimal hash, and use that to reduce
+    /// the amount of work needed to produce the hash.
+    /// </summary>
+    private static async ValueTask<(GamePath gamePath, Hash newHash)> HashGameFile(IDb hashDb, GameInstallation installation, AbsolutePath file, CancellationToken token)
+    {
+        var gamePath = installation.LocationsRegister.ToGamePath(file);
+        
+        // It's cheapest to look at the size first, if there's no matching size then we don't do a minimal hash
+        var relation = HashRelation.FindBySize(hashDb, file.FileInfo.Size);
+        if (relation.Count != 0)
+        {
+            // We have a size match, so minimal hash it, and see if we have a match
+            var minimalHash = await MultiHasher.MinimalHash(file, token);
+            foreach (var matchingHash in HashRelation.FindByMinimalHash(hashDb, minimalHash))
+            {
+                // Match on the relative path as well as the minimal hash
+                foreach (var path in matchingHash.Paths)
+                {
+                    if (path.Path == gamePath.Path)
+                        return (gamePath, matchingHash.XxHash3);
+                }
+                
+            }
+        }
```

Comment:
It seems this code is making two DB lookups in case there are size matches.
Wouldn't it be more efficient to only lookup the hash matches in the already filtered relations that match by size?

I'm not sure about the performance numbers, but my instincts would be to:
- lookup by size
- filter by matching paths
- perform minimal hash only if a match is found

I guess this might get better with better query syntax anyways in the not far future

Reply:
I commented it a bit, the chances for collisions here are rare, and we can improve this in the future if we need to. For now I think it'll run pretty fast.

Reply:
The code is doing the hash lookup on the entire DB, should it not only check the Size matches? Or does it not matter due to const lookup times?

Reply:
It doesn't matter here so much because all we're doing is finding a relation between a `size, path, minimal hash` and a `xxhash3`. We will be much more restrictive when we start building loadouts. In otherwords, a minimal hash in a given place, for any game is unlikely to collide with any other minimal hash in the same place from a different game.